catroid/src/main/java/org/catrobat/catroid/formulaeditor/datacontainer/ProjectDataBehaviour.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.formulaeditor.datacontainer;
+
+import java.util.List;
+
+abstract class ProjectDataBehaviour<V> {
+
+	protected abstract List<V> getDataList();
+	protected abstract V newInstance(String name);
+	protected abstract String getDataName(V data);
+	protected abstract void setDataName(V data, String name);
+
+	V add(String name) {
+		V dataToAdd = newInstance(name);
+		getDataList().add(dataToAdd);
+		return dataToAdd;
+	}
+
+	V rename(String newName, String oldName) {
+		V dataToRename = find(getDataList(), oldName);
+		setDataName(dataToRename, newName);
+		return dataToRename;
+	}
+
+	boolean exists(String name) {
+		for (V data : getDataList()) {
+			if (getDataName(data).equals(name)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private V find(List<V> dataList, String name) {
+		if (dataList == null) {
```

Comment:
just checked it, nope couldn't as a new ArrayList is created if the project lists or variables are null in getProjectLists() and getProjectVariables() in Project.
I've removed that check now.